internal/controllers/addon_controller.go

```diff
@@ -27,3 +58,90 @@ func (r *AddonReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&addonv1alpha1.Addon{}).
 		Complete(r)
 }
+
+func (r *AddonReconciler) ensureDeletionOfUnwantedNamespaces(
+	ctx context.Context, addon *addonv1alpha1.Addon) error {
+	// steps:
+	// - get all namespaces that are managed by this controller + this addon (in other words: have common labels)
+	// - remove all namespaces that are no longer in addon.Spec.Namespaces
+	// - report success/failure condition
+
+	// notes:
+	// - separate condition?
+	// - separate phase?
+	return nil
+}
+
+func (r *AddonReconciler) ensureWantedNamespaces(
+	ctx context.Context, addon *addonv1alpha1.Addon) error {
+	var ensuredNamespaces []*corev1.Namespace
+
+	// how would I reflect a namespace name collision in addon.Status?
```

Comment:
I'd not mention collisions in the status.

I was thinking about that today. We can play nice here and just allow collisions. We then only remove the namespace when it's not specified by any addon.

Think of the namespaces list as "this addon needs these namespaces", not as "these are my namespaces, don't touch them".

Anyway, not sure you're considering to fail on collisions here, but I'd consider failing here as a bad thing, a bit too late. If we want to do semantic validation, we should consider admission webhooks and use them also for gatekeeping, avoiding offending CRs to get to the clusters.

Reply:
Having multiple owners of an object (e.g. n*Addon objects owning n*Namespaces) will make this PR way more complicated.
If it's not a required feature, we should not complicate our life's before we have to.

Webhooks are great for in-depth validation, but doing validation in context of the current cluster state is generally a bad idea. The kubernetes way is usually to deny object creation if the object is syntactically, semantically (CRD Scheme) or logically (Webhook) wrong. 
While everything transient, like missing configuration or namespaces is not bound to a particular order and will be retried until success, while reporting issues until resolved.

Even if we disregard Kubernetes API sementics, we have two big technical issues:
- Asserting on other objects is racy. Namespaces might be created or deleted after the webhook ran through. So we have to deal with exceptions here anyway.
- If we go through with putting preflight checks into the webhook, we would essentially duplicate half of what our controller(s) are doing into the webhook.
- We need a "human controller flow" on OCM-Level to retry creating the Addon when it's denied in admission.

I would suggest to report a `NamespaceCollision` condition on the Addon CR and stopping executing further steps in that namespace, as we don't know who is controlling that namespace and thus cannot manage it.
We can also use an EventRecorder to post a new event on the Addon CR to inform admins everytime we check and run into this issue, while the Condition tracks since when the issue is reported.

Reply:
Ok, all good points. I think we are heading to a full blown cluster for gatekeeping also, which is probably a nice to have anyway.